# Mu-Pro_manuscript

**Table S1. 35 *Mutator* mutant alleles isolated from the UniformMu population in maize.** Mutant allele transcriptome data obtained, transcript abundance, and transcriptome assembly predicted transcript structure. RNAseq CPM and FPKM for mutant and wild type W22, DE data averaged across biological replicates (N), log2fc: log2 fold change of mutant to control, lfcSE: log fold change standard error and FDR adjusted p-value.

**Table S2. Gene expression values for 24 transcription factor genes in different tissues** The expression value (CPM) for each of the 22 TFs was assessed based on prior sampling of tissues or developmental stages in B73 (Zhou et al. 2019). Values highlighted indicate the predicted expression level of TF genes in tissues sampled for RNAseq in this study.

**Table S3. Mutant allele *Mu* element identity and orientation by gDNA PCR.** Table follows the format of Figure S2 with primer sets not tested grayed out, and tested primer sets resulting in amplification—blue, no amplification—pink.

**Table S4. Mutant allele transcript boundaries and potential for *Mu* read-through tested by RT-PCR.** Transcript boundaries of gene promoter-*Mu* and *Mu* promoter transcripts: Table follows format of Figure 5B with primer *Mu*-specific primer names listed above the *Mu* sequence amplified and the gene-specific primer used for each allele in the corresponding row. Some *Mu*-specific
primers have specificity to both *Mu* TIRs and can be used with both gene-specific primers to determine transcript boundaries. The four alleles tested with gene-specific primers flanking *Mu* are included.

**Table S5. Transcript abundance for shared exon sequence between mutant and wild type transcripts.** Genomic coordinates for exons with shared sequence between mutant allele transcripts: gene promoter partial or *Mu* promoter and the corresponding wild type transcript(s) and RNAseq read counts (count) for each exon are included. Library effect size was calculated for each sample. Transcript abundance is reported as CPM per fragment--see Methods. Genomic coordinates for shared exons may not be complete annotated exons. 

**Table S6. Tissue-specific expression patterns for mutant and wild type transcripts tested by RT-qPCR.** Delta Ct for each replicate and priemrs used are included.



**Data File S1.** FASTA file of TRINITY *de novo* assembled transcripts for mutant and wild type alleles. 

**Data File S2.** Mutant Allele PCR, RT-PCR, and RT-qPCR gene-specific and Mu-specific primers.
