# Mu Promoter Manuscript

**Table S1. 35 *Mutator* mutant alleles isolated from the UniformMu population in maize.** Mutant allele transcriptome data obtained, transcript abundance, and transcriptome assembly predicted transcript structure. UniformMu Mu ID and Stock, Pos: *Mu* insertion position within the gene, genotyping primers used to isolate the homozygous mutant, and Pedi: pedigree of the mutant stock sampled for RNA-seq (BC = backcross, S = self). RNA-seq CPM and FPKM for mutant and wild-type W22, DE data averaged across biological replicates (N), log2fc: log2 fold change of mutant to control, lfcSE: log fold change standard error and FDR adjusted p-value. Data for TF genes that are that are DE in the mutant compared to the wild-type are highlighted. Transcript assembly transcript structure references categories described in Figure 3A.

**Table S2. Gene expression values for 24 transcription factor genes in different tissues.** The expression value (CPM) for each of the 22 TFs was assessed based on prior sampling of tissues or developmental stages in B73 (Zhou et al. 2019). Values highlighted indicate the predicted expression level of TF genes in tissues sampled for RNA-seq in this study.

**Table S3. Mutant allele *Mu* element identity and orientation by gDNA PCR.** Table follows the format of Figure S2 with primer sets not tested-gray, and tested primer sets resulting in amplification—blue, no amplification—pink.

**Table S4. Mutant allele transcript boundaries and potential for *Mu* read-through tested by RT-PCR.** 
Transcript boundaries of gene TSS-*Mu* and *Mu* TSS transcripts: Table follows format of Figure 5B with *Mu*-specific primers listed above the *Mu* sequence amplified (bp) and the gene-specific primer used for each allele in the corresponding row. Tan; RT-PCR amplification, Pink; absence of RT-PCR amplification and the region where the transcript terminates, Black; absence of RT-PCR amplification. Some *Mu*-specific primers used have specificity to both 5’ and 3’ *Mu* TIRs. The three alleles tested with gene-specific primers flanking *Mu* are included (Figure 2C).

**Table S5. Transcript abundance for shared exon sequence between mutant and wild-type transcripts.** Counts per million (CPM) per fragment calculated for each mutant allele transcript, gene TSS partial or *Mu* TSS, and the corresponding wild type W22 transcript(s) is shown—see Methods for calculation and normalization. The distance in bp of the Mu insertion from the annotated W22 TSS is listed: bp_TSS. Gene TSS refers to gene TSS partial or gene TSS-*Mu* transcripts.

**Table S6. Tissue-specific expression patterns for mutant and wild-type transcripts tested by RT-qPCR.** The average delta Ct +/- standard deviation for biological replicates of each mutant allele and W22 transcript in the 6 tissues tested: coleoptile tip, root, shoot, flag leaf, ear spikelet and tassel stem. RT-qPCR primers used are listed: F_ID and R_ID.

**Data File S1.** FASTA file of TRINITY *de novo* assembled transcripts for mutant and wild type alleles. **Data File S1 README** file included for additional notes on transript assemblies.

**Data File S2.** Mutant Allele PCR, RT-PCR, and RT-qPCR gene-specific and *Mu*-specific primers.

**Note:** Tables with raw data are listed as **_raw** and tables included in the manuscript supplemental tables file are listed as **_ms**. RNA-seq data collected can be found at SRA BioProject: SUB12766103 and RNA-seq meta data is included as **Table_SX_RNAseq_meta-data.**
